p4_symbolic/symbolic/symbolic.h

```diff
@@ -0,0 +1,114 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef P4_SYMBOLIC_SYMBOLIC_SYMBOLIC_H_
+#define P4_SYMBOLIC_SYMBOLIC_SYMBOLIC_H_
+
+#include <map>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "absl/status/status.h"
+#include "p4_pdpi/utils/status_utils.h"
+#include "p4_symbolic/ir/ir.pb.h"
+#include "z3++.h"  // added as a system dependency for now.
+
+namespace p4_symbolic {
+namespace symbolic {
+
+class Analyzer {
+ protected:
+  ir::P4Program kProgram;
+  z3::context kContext;
+
+  // Maps header fields full names to their corresponding symbolic expression.
+  // A header field name is on the form <header_type_name>.<field_name>.
+  std::unordered_map<std::string, z3::expr> kFieldsMap;
+
+  // Maps variable full names to their corresponding symbolic expression.
+  // A variable full name is on the form <action_full_name>.<variable_name>,
+  // where action is the enclosing action in which the variable was defined.
+  std::unordered_map<std::string, z3::expr> kVariablesMap;
+
+  // Maps a table full name and row index to the corresponding symbolic
+  // expression representing matching with that row, such that:
+  // entries_map['table1_3'] <=> row at index 3 in table1 was hit.
+  std::unordered_map<std::string, z3::expr> kEntriesMap;
+
+  // Maps an action full name to the corresponding symbolic expressions
+  // representing that this action was indeed matched on and called.
+  std::unordered_map<std::string, z3::expr> kActionsMap;
+
+  // Symbolic constraints that describe relationships between various variables
+  // and fields in the program.
+  std::vector<z3::expr> kConstraints;
```

Comment:
Instead of creating constraints, I think we should write the code to create symbolic expressions, and only at the very end do we assert something. In particular, the entire switch pipeline should be a symbolic expression. The expression should have the access to the input packet, and should produce a trace with all the information about how the packet was forwarded (which table entry was hit, was the packet dropped, etc). You can look at the code that I shared with you for more detail. The user can then write an assertion over that expression, e.g. packet must not have dropped, and must have hit table entry FOO. 

If you follow that style, your program will look very nice, because it looks just like a non-symbolic P4 simulator that happens to operate with symbolic values and operations.